### PR TITLE
improve dependecies/remove noarch

### PR DESCRIPTION
**Pull request by carlodri**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Not sure if this is a good idea, or even if it is allowed to have a mix of `noarch` and non-`noarch` outputs....

Opening in the attempt to address issues like #44, but, again, not sure if it is the right approach also in terms of costs/benefit.

Closes #44

### CODE REVIEW

**Comment by carlodri** [comment]
@conda-forge-admin please rerender

**Comment by carlodri** [comment]
something weird going on... selectors for `typed-ast` seem not to work properly?

**Comment by carlodri** [comment]
@conda-forge/black let me know your thoughts on this (and on the selector not working, if you have a hint)

**Comment by deantvv** [comment]
@carlodri It seems like conda-build is smart enough to drop the `typed-ast` if it isn't actually used in build. After making a few [changes](https://github.com/conda-forge/black-feedstock/commit/bfd1af8f2fc47cd78b85bd19b6c20a149843f5d0), it seems to build and passes tests on my local environment.

**Comment by carlodri** [comment]
@conda-forge-admin please rerender

**Comment by carlodri** [comment]
> @carlodri It seems like conda-build is smart enough to drop the `typed-ast` if it isn't actually used in build. After making a few [changes](https://github.com/conda-forge/black-feedstock/commit/bfd1af8f2fc47cd78b85bd19b6c20a149843f5d0), it seems to build and passes tests on my local environment.

I'm not sure it is a good idea to release all constraints on the `pin_subpackage`... but I'm really not sure of what it does exactly.

**Comment by deantvv** [comment]
We need to release at least `exact=True` for `pin_subpackage`  because when `exact=True` it will include build variant which won't work in this case because only `black` is built for different variants but other packages are built with `noarch: python`.

We also need following diff to make conda-build include `typed-ast` when building that way it will include `typed-ast` in dependencies.
```diff
diff --git a/recipe/meta.yaml b/recipe/meta.yaml
index 3b071d6..0bd2f32 100644
--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -24,6 +24,7 @@ outputs:
         - python
         - pip
         - setuptools_scm
+        - typed-ast >=1.4.2  # [py<38 and python_impl == 'cpython']
       run:
         - python
         - click >=8.0.0
```

**Comment by carlodri** [comment]
@conda-forge-admin please rerender

**Comment by carlodri** [comment]
> We need to release at least `exact=True` for `pin_subpackage` because when `exact=True` it will include build variant which won't work in this case because only `black` is built for different variants but other packages are built with `noarch: python`.

thanks for the insight! But shouldn't we constrain it at least to the same `x.x.x` version as in the black recipe output? As you can see I tried using `max_pin` but doesn't seem to help...

**Comment by deantvv** [comment]
> thanks for the insight! 

It turns out that my "insight" is wrong. The reason why `typed-ast` kept getting drop is because `python_impl == 'cpython'` is False so the whole line is removed. So I start looking why there isn't `python_impl` in our build. It turns out that we only support `python_impl=='cpython'` so rerender never add `python_impl: cpython` to the build. We could drop `python_impl == 'cpython'` for now and hopefully the CI will turn green.

> To build your python package for pypy, wait for the bot to send a PR and contact conda-forge/bot team if a PR is not sent after the dependencies have been built. [conda-forge document](https://conda-forge.org/docs/maintainer/knowledge_base.html)

If we want to support `pypy` we will need to have all dependencies available in either noarch or pypy build. Now we still have `click` and `tomli` only available in `cpython` build.

> But shouldn't we constrain it at least to the same x.x.x version as in the black recipe output? As you can see I tried using max_pin but doesn't seem to help...

I didn't see why you think it isn't the case. In my local build, it constrain to `>=x.x.x,<x.x.x+1`

File: recipe/meta.yaml [hunk 1]
```diff
@@ -10,30 +10,31 @@ source:
...
+      skip: true  # [py<36]
       entry_points:
         - black = black:patched_main
     requirements:
       host:
-        - python >=3.6
+        - python
         - pip
         - setuptools_scm
+        - typed-ast >=1.4.2  # [py<38 and python_impl == 'cpython']
```
Comment (deantvv) [inline recipe/meta.yaml:27]: My "insight" is wrong, so please remove this line if you would.
```suggestion
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -10,30 +10,31 @@ source:
...
       run:
-        - python >=3.6
-        - platformdirs >=2
+        - python
         - click >=8.0.0
-        - dataclasses >=0.6
+        - platformdirs >=2
+        - tomli >=1.1.0  # [py<311]
+        - typed-ast >=1.4.2  # [py<38 and python_impl == 'cpython']
```
Comment (deantvv) [inline recipe/meta.yaml:33]:
```suggestion
        + typed-ast >=1.4.2  # [py<38]
```

**Review by deantvv — CHANGES_REQUESTED** [review]
> thanks for the insight! 

It turns out that my "insight" is wrong. The reason why `typed-ast` kept getting drop is because `python_impl == 'cpython'` is False so the whole line is removed. So I start looking why there isn't `python_impl` in our build. It turns out that we only support `python_impl=='cpython'` so rerender never add `python_impl: cpython` to the build. We could drop `python_impl == 'cpython'` for now and hopefully the CI will turn green.

> To build your python package for pypy, wait for the bot to send a PR and contact conda-forge/bot team if a PR is not sent after the dependencies have been built. [conda-forge document](https://conda-forge.org/docs/maintainer/knowledge_base.html)

If we want to support `pypy` we will need to have all dependencies available in either noarch or pypy build. Now we still have `click` and `tomli` only available in `cpython` build.

> But shouldn't we constrain it at least to the same x.x.x version as in the black recipe output? As you can see I tried using max_pin but doesn't seem to help...

I didn't see why you think it isn't the case. In my local build, it constrain to `>=x.x.x,<x.x.x+1`

**Comment by carlodri** [comment]
@conda-forge-admin please rerender

**Pull request merged**